django/db/models/fields/composite.py

```diff
@@ -0,0 +1,415 @@
+import copy
+from collections import OrderedDict
+from itertools import chain
+import re
+import weakref
+
+from django.core import checks
+from django.core.exceptions import FieldDoesNotExist
+from django.db.models.fields import Field
+from django.db.models.fields.composite_lookups import (
+    CompositeExact, CompositeIn, CompositeIsNull, SubfieldTransform
+)
+from django.utils import six
+from django.utils.functional import cached_property
+from django.utils.observables import Observable, ObservableDict
+
+
+__all__ = ('constrain', 'CompositeField')
+
+
+def constrain(*fields, **kwargs):
```

Comment:
Marc started some work on class-based indexes: https://github.com/django/django/pull/3253. I think that might solve the use case here?